src/VectorBundle.jl

```diff
@@ -0,0 +1,221 @@
+end
+
+function zero_vector!(M::Manifold, VS::TCoTSpaceType, v, x)
+    zero_tangent_vector!(M, v, x)
+    return v
+end
+
+"""
+    zero_vector(M::Manifold, VS::VectorSpaceType, x)
+
+Compute the zero vector from the vector space of type `VS` at point `x`
+from manifold `M`.
+"""
+function zero_vector(M::Manifold, VS::VectorSpaceType, x)
+    v = similar_result(M, VS, zero_vector, x)
+    zero_vector!(M, VS, v, x)
+    return v
+end
+
+"""
+    project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+"""
+    inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+
+Inner product of vectors `v` and `w` from the vector space of type `VS`
+at point `x` from manifold `M`.
+"""
+function inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+    error("inner not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)), point of type $(typeof(x)) and vectors of types $(typeof(v)) and $(typeof(w)).")
+end
+
+function inner(M::Manifold, VS::TCoTSpaceType, x, v, w)
+    return inner(M, x, v, w)
+end
+
+function project_vector!(M::Manifold, VS::TCoTSpaceType, v, x, w)
+    project_tangent!(M, v, x, w)
+    return v
+end
+
+"""
+    distance(M::Manifold, VS::VectorSpaceType, x, y)
+"""
+function distance(M::Manifold, VS::VectorSpaceType, x, y)
+    error("distance not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)) and vectors of types $(typeof(x)) and $(typeof(y)).")
+end
+
+distance(M::Manifold, VS::TCoTSpaceType, x, y) = norm(x-y)
+
+
+"""
+    norm(M::Manifold, VS::VectorSpaceType, x, v)
+
+Norm of the vector `v` from the vector space of type `VS` at point `x`
+from manifold `M`.
+"""
+function norm(M::Manifold, VS::VectorSpaceType, x, v)
+    error("norm not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)) and vectors of types $(typeof(x)) and $(typeof(v)).")
+end
+
+norm(M::Manifold, VS::TCoTSpaceType, x, v) = norm(v)
+
+"""
+    VectorBundle(M::Manifold, type::VectorSpaceType)
+
+Vector bundle on manifold `M` of type `type`.
+"""
+struct VectorBundle{TVS<:VectorSpaceType, TM<:Manifold} <: Manifold
+    type::TVS
+    M::TM
+end
+
+function representation_size(M::VectorBundle)
+    len_manifold = prod(representation_size(M.M))
+    len_vs = prod(representation_size(M.M, M.type))
+    return (len_manifold + len_vs,)
+end
+
+TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)
+CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
```

Comment:
Would it be better to declare these as types too for dispatch? e.g.

```julia
TangentBundle{M} = VectorBundle{TangentSpace,M}
CotangentBundle{M} = VectorBundle{CotangentSpace,M}
foo(T::TangentBundle) = ...
```

Reply:
I've tried that and it looks like the name `TangentBundle` can refer either to the function I've defined or be an alias for `VectorBundle{TangentSpaceType, M}`, so one of these things needs a different name.

Reply:
This worked for me

```julia
TangentBundle{M} = VectorBundle{TangentSpaceType,M}
TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)

CotangentBundle{M} = VectorBundle{CotangentSpaceType,M}
CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
```

```julia
julia> TM = Manifolds.TangentBundle(Sphere(2))
VectorBundle{Manifolds.TangentSpaceType,Sphere{2}}(Manifolds.TangentSpaceType(), Sphere{2}())

julia> typeof(TM)<:Manifolds.TangentBundle
true
```

Reply:
Thanks, it doesn't work if the lines are in the reverse order and I didn't think of trying it this way.

Reply:
Just something for discussion: Why is `M` the second type argument of `VectorBundle`? I would've thought it'll e the first (design-wise) Since I would read it as “VectorBundle over M of X-type”

Reply:
`M` is second because I think it's more common to dispatch on the vector space type than on the manifold, so we can write for example `VectorBundle{TangentSpaceType}` instead of `VectorBundle{<:Manifold, TangentSpaceType}`. I generally try to order type arguments from most likely to dispatch on to least likely to dispatch on.